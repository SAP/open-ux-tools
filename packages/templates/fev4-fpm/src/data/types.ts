
export interface CustomPage {
    navigation: {
        sourcePage: string;
        targetEntity: string;
    },
    view: {
        name?: string;
        title?: string;
        path?: string;
    }
}

export interface CustomPageConfig extends CustomPage {
    app: {
        id: string;
    }
}
