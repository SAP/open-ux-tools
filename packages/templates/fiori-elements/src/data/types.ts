import { Ui5App } from '@sap/ux-ui5-application-template';
import { OdataService } from '@sap/ux-odata-service-template';

export enum TemplateType {
    ListReport = 'lrop',
    Form = 'form',
    OverviewPage = 'ovp'
}

export enum FrameworkVersion {
    V2 = 'v2',
    V4 = 'V4'
}

export interface Template<T = {}> {
    type: TemplateType;
    version: FrameworkVersion;
    settings: T;
}

export interface LROP {
    mainEntity: string;
    navigationEntity?: string;
}
export interface OVP {
    filterEntity: string;
}
export interface FEApp<T> extends Ui5App {
    template: Template<T>;
    annotations: {
        ns?: string;
    };
    service: OdataService;
}
