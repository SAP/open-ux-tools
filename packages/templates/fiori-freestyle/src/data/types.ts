import { Ui5App } from '@sap/ux-ui5-application-template';
import { OdataService } from '@sap/ux-odata-service-template';

export enum TemplateType {
    Basic = 'basic',
    Worklist = 'worklist',
    ListDetail = 'listdetail'
}

interface Entity {
    name: string;
    key: string;
    idProperty: string;
    numberProperty?: string;
    unitOfMeasureProperty?: string;
}

export interface WorklistSettings {
    entity: Entity;
}

export interface ListDetailSettings {
    mainObject: Entity;
    lineItem: Entity;
}

export interface Template<T = {}> {
    type: TemplateType;
    settings: T;
}

export interface FreestyleApp<T> extends Ui5App {
    template: Template<T>;
    service?: OdataService;
}