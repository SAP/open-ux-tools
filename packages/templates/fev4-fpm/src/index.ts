import { join } from 'path';
import { create as createStorage } from 'mem-fs';
import { create, Editor } from 'mem-fs-editor';
import { render } from 'ejs';

import { CustomPage, CustomPageConfig, enhanceData } from './data';

/**
 * Validates the provided base path.
 *
 * @param {string} basePath - the base path
 * @param {Editor} fs - the memfs editor instance
 */
function validateBasePath(basePath: string, fs: Editor) {
    [join(basePath, 'webapp', 'manifest.json')].forEach(
        (path) => {
            if (!fs.exists(path)) {
                throw new Error(`Invalid project folder. Cannot find required file ${path}`);
            }
        }
    );
}
/**
 * Writes the template to the memfs editor instance.
 *
 * @param {string} basePath - the base path
 * @param {CustomPage} data - the custom page configuration
 * @param {Editor} [fs] - the memfs editor instance
 * @returns {Promise<Editor>} the updated memfs editor instance
 */
async function generateCustomPage(basePath: string, data: CustomPage, fs?: Editor): Promise<Editor> {
    if (!fs) {
        fs = create(createStorage());
    }
    validateBasePath(basePath, fs);
    
    
    const manifestPath = join(basePath, 'webapp/manifest.json');
    const config = enhanceData(data, manifestPath, fs);
    
    // merge content into existing files
    const root = join(__dirname, '..', 'templates/customPage');

    // enhance manifest.json
    fs.extendJSON(manifestPath, JSON.parse(render(fs.read(join(root, `manifest.json`)), config)));

    // add extension content
    if (data.view.path === undefined) {
        data.view.path = join(basePath, `webapp/ext/view/${config.view.name}.view.xml`);
        fs.copyTpl(
            join(root,'ext/view/CustomPage.view.xml'), 
            data.view.path, config);
        fs.copyTpl(
            join(root,'ext/controller/CustomPage.controller.js'), 
            join(basePath, `webapp/ext/controller/${config.view.name}.controller.js`), config);
    }

    return fs;
}

export { CustomPage, generateCustomPage};
