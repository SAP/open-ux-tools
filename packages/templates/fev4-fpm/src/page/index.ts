import { join, dirname } from 'path';
import { create as createStorage } from 'mem-fs';
import { create, Editor } from 'mem-fs-editor';
import { render } from 'ejs';

import { CustomPage, CustomPageConfig } from '../types';

/**
 * Enhances the provided custom page configuration with default data
 *
 * @param {CustomPage} data - a custom page configuration object
 */
 export function enhanceData(data: CustomPage, manifestPath: string, fs: Editor): CustomPageConfig {
    // enforce naming conventions
    const firstChar = data.name[0];
    const nameForId = firstChar.toLocaleLowerCase() + data.name.substring(1);
    data.name = firstChar.toUpperCase() + data.name.substring(1)

    const manifest: any = fs.readJSON(manifestPath);
    const config: CustomPageConfig = {
        ...data,
        id: `${manifest['sap.app'].id}.ext.${nameForId}`,
        path: join(dirname(manifestPath), 'ext', nameForId)
    };
    if (config.view === undefined) {
        config.view = {
            title: config.name
        };
    }
    return config;
}

/**
 * Validates the provided base path.
 *
 * @param {string} basePath - the base path
 * @param {Editor} fs - the memfs editor instance
 */
export function validateBasePath(basePath: string, fs?: Editor) {
    if (!fs) {
        fs = create(createStorage());
    }

    const manifestPath = join(basePath, 'webapp', 'manifest.json');
    if (!fs.exists(manifestPath)) {
        throw new Error(`Invalid project folder. Cannot find required file ${manifestPath}`);
    } else {
        const manifest = fs.readJSON(manifestPath);
        // TODO: check if this is an FE app
    }

}
/**
 * Writes the template to the memfs editor instance.
 *
 * @param {string} basePath - the base path
 * @param {CustomPage} data - the custom page configuration
 * @param {Editor} [fs] - the memfs editor instance
 * @returns {Promise<Editor>} the updated memfs editor instance
 */
export async function generateCustomPage(basePath: string, data: CustomPage, fs?: Editor): Promise<Editor> {
    if (!fs) {
        fs = create(createStorage());
    }
    validateBasePath(basePath, fs);
    
    
    const manifestPath = join(basePath, 'webapp/manifest.json');
    const config = enhanceData(data, manifestPath, fs);
    
    // merge content into existing files
    const root = join(__dirname, '../../templates/page');

    // enhance manifest.json
    fs.extendJSON(manifestPath, JSON.parse(render(fs.read(join(root, `manifest.json`)), config)), ( key, value ) => {
        if (key === 'routes') {
            const routes = value as object[];
            routes.push({
                pattern: `${config.navigation ? config.navigation.sourceEntity + '({key})/' + config.navigation.navEntity + '({key2})' : config.entity + '({key})'}:?query:`,
                name: `${config.entity}${config.name}`,
                target: `${config.entity}${config.name}`
            });
        }
        return value;
    });

    // add extension content
    if (config.view?.path) {
        // TODO: copying is not that simple, controller path needs to be read from view.xml and if the controller is copied, it's id in the sources needs to be changed
        fs.copy(config.view.path, join(config.path, `${config.name}.view.xml`));
        fs.copy(config.view.path.replace('view.xml', 'controller.js'), join(config.path, `${config.name}.controller.js`));
    } else {
        fs.copyTpl(
            join(root,'ext/NAME/View.xml'), 
            join(config.path, `${config.name}.view.xml`), config);
        fs.copyTpl(
            join(root,'ext/NAME/Controller.js'), 
            join(config.path, `${config.name}.controller.js`), config);
    }

    return fs;
}