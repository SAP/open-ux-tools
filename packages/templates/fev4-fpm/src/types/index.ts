export * from './page';

export interface Ui5Route {
    name: string;
    pattern: string;
    target: string | string[];
}
