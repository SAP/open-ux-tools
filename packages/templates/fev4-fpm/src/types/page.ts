
export interface CustomPage {
    name: string;
    entity: string;
    navigation?: {
        sourcePage?: string;
        targetEntity?: string;
    },
    view?: {
        title?: string;
        path?: string;
    }
}

export interface CustomPageConfig extends CustomPage {
    id: string;
    path: string;
}
