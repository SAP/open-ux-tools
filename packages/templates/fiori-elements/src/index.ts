import { join } from 'path';
import { Editor } from 'mem-fs-editor';
import { render } from 'ejs';

import { generate as generateUi5Project } from '@sap/ux-ui5-application-template';
import { generate as addOdataService } from '@sap/ux-odata-service-template';
import { FEApp, getBaseComponent } from './data';
import { UI5Config } from '../../../lib/ui5-config/dist';
import { getUI5Libs } from './data/ui5Libs';
import { getMiddlewareConfig } from './data/middlewareConfig';

/**
 * @param basePath
 * @param data
 * @param fs
 */
async function generate<T>(basePath: string, data: FEApp<T>, fs?: Editor): Promise<Editor> {
    // generate base UI5 project
    data.app.baseComponent = getBaseComponent(data.template);
    fs = await generateUi5Project(basePath, data, fs);

    // add new and overwrite files from templates e.g. annotations.xml
    const tmpPath = join(__dirname, '..', 'templates', 'add');
    fs.copyTpl(join(tmpPath, '**/*.*'), basePath, data);

    // merge content into existing files
    const extRoot = join(__dirname, '..', 'templates', 'extend');

    // package.json
    const packagePath = join(basePath, 'package.json');
    fs.extendJSON(packagePath, fs.readJSON(join(extRoot, 'package.json')));
    const packageJson = JSON.parse(fs.read(packagePath));
    packageJson.ui5.dependencies.push('@sap/ux-ui5-tooling');
    fs.writeJSON(packagePath, packageJson);

    // manifest.json
    const manifestPath = join(basePath, 'webapp', 'manifest.json');
    fs.extendJSON(manifestPath, fs.readJSON(join(extRoot, 'manifest.json')));
    fs.extendJSON(
        manifestPath,
        JSON.parse(render(fs.read(join(extRoot, data.template.version, `manifest.${data.template.type}.json`)), data))
    );

    // ui5.yaml
    const ui5ConfigPath = join(basePath, 'ui5.yaml');
    const ui5Config = await UI5Config.newInstance(fs.read(ui5ConfigPath));
    ui5Config
        .addLibraries(getUI5Libs(data.template.type, data.template.version))
        .addCustomMiddleware(getMiddlewareConfig());
    fs.write(ui5ConfigPath, ui5Config.toString());

    // add service to the project
    await addOdataService(basePath, data.service, fs);
    const manifest = JSON.parse(fs.read(manifestPath));
    manifest['sap.app'].dataSources[data.service.name!].settings.annotations.push('annotation');
    fs.writeJSON(manifestPath, manifest);
    return fs;
}

export { generate };
