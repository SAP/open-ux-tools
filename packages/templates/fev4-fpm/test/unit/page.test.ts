import { create as createStorage } from 'mem-fs';
import { create, Editor } from 'mem-fs-editor';
import { join } from 'path';
import { generateCustomPage } from '../../src';
import { Ui5Route } from '../../src/types';

describe('CustomPage', () => {
    let fs = create(createStorage());
    const testDir = 'virtual-temp';
    const testAppManifest = JSON.stringify({
        "sap.app": {
            id: "my.test.App"
        },
        "sap.ui5": {
            routing: {
                routes: [] as Ui5Route[]
            },
            targets: {
                TestObjectPage: {

                }
            }
        }
    }, null, 2);

    beforeAll(() => {
        fs.delete(testDir);
    });

    test('Add a custom page with minimal input', async () => {
        const target = join(testDir, '' + Date.now());
        fs.write(join(target, 'webapp/manifest.json'), testAppManifest);
        generateCustomPage(target, {
            name: "CustomPage",
            entity: "RootEnity"
        }, fs);
        expect((fs as any).dump(target)).toMatchSnapshot();
    });

    const inputWithNavigation = {
        name: "CustomPage",
        entity: "ChildEntity",
        navigation: {
            sourcePage: "TestObjectPage",
            sourceEntity: "RootEntity",
            navEntity: "navToChildEntity"
        }
    };

    test('Add a custom page with navigation to it', async () => {
        const target = join(testDir, '' + Date.now());
        fs.write(join(target, 'webapp/manifest.json'), testAppManifest);
        generateCustomPage(target, inputWithNavigation, fs);
        expect(fs.readJSON(join(target, 'webapp/manifest.json'))).toMatchSnapshot(); 
    });

    test('Add a custom page to app with target as array', async () => {
        const testManifestWithArray = JSON.parse(testAppManifest);
        testManifestWithArray['sap.ui5'].routing.routes.push({
            pattern: "object/{key}",
            name: "TestObjectPage",
            target: [
                "TestObjectPage"
            ]
        });
        const target = join(testDir, '' + Date.now());
        fs.writeJSON(join(target, 'webapp/manifest.json'), testManifestWithArray);
        generateCustomPage(target, inputWithNavigation, fs);
        expect(fs.readJSON(join(target, 'webapp/manifest.json'))).toMatchSnapshot(); 
    });
});
