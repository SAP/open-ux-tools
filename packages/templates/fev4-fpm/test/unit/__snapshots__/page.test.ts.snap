// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomPage Add a custom page to app with target as array 1`] = `
Object {
  "sap.app": Object {
    "id": "my.test.App",
  },
  "sap.ui5": Object {
    "routing": Object {
      "routes": Array [
        Object {
          "name": "TestObjectPage",
          "pattern": "object/{key}",
          "target": Array [
            "TestObjectPage",
            "ChildEntityCustomPage",
          ],
        },
        Object {
          "name": "ChildEntityCustomPage",
          "pattern": "RootEntity({key})/navToChildEntity({key2}):?query:",
          "target": Array [
            "TestObjectPage",
            "ChildEntityCustomPage",
          ],
        },
      ],
      "targets": Object {
        "ChildEntityCustomPage": Object {
          "id": "my.test.App.ext.customPage",
          "name": "sap.fe.core.fpm",
          "options": Object {
            "settings": Object {
              "entitySet": "ChildEntity",
              "navigation": Object {},
              "viewName": "my.test.App.ext.customPage.CustomPage",
            },
          },
          "type": "Component",
        },
        "TestObjectPage": Object {
          "options": Object {
            "settings": Object {
              "navigation": Object {
                "navToChildEntity": Object {
                  "detail": Object {
                    "route": "ChildEntityCustomPage",
                  },
                },
              },
            },
          },
        },
      },
    },
    "targets": Object {
      "TestObjectPage": Object {},
    },
  },
}
`;

exports[`CustomPage Add a custom page with minimal input 1`] = `
Object {
  "webapp/ext/customPage/CustomPage.controller.js": Object {
    "contents": "sap.ui.define(
    [
        'sap/fe/core/PageController',
        'sap/fe/core/controllerextensions/RoutingListener'
    ],
    function(Controller, RoutingListener) {
        'use strict';

        return Controller.extend('my.test.App.ext.customPage.CustomPage', {
            routingListener: RoutingListener,

            /**
             * Called when a controller is instantiated and its View controls (if available) are already created.
             * Can be used to modify the View before it is displayed, to bind event handlers and do other one-time initialization.
             * @memberOf my.test.App.ext.customPage.CustomPage
             */
            //	onInit: function () {
            //
            //	},

            /**
             * Similar to onAfterRendering, but this hook is invoked before the controller's View is re-rendered
             * (NOT before the first rendering! onInit() is used for that one!).
             * @memberOf my.test.App.ext.customPage.CustomPage
             */
            //	onBeforeRendering: function() {
            //
            //	},

            /**
             * Called when the View has been rendered (so its HTML is part of the document). Post-rendering manipulations of the HTML could be done here.
             * This hook is the same one that SAPUI5 controls get after being rendered.
             * @memberOf my.test.App.ext.customPage.CustomPage
             */
            //	onAfterRendering: function() {
            //
            //	},

            /**
             * Called when the Controller is destroyed. Use this one to free resources and finalize activities.
             * @memberOf my.test.App.ext.customPage.CustomPage
             */
            //	onExit: function() {
            //
            //	}
        });
    }
);
",
    "state": "modified",
  },
  "webapp/ext/customPage/CustomPage.view.xml": Object {
    "contents": "<mvc:View xmlns:core=\\"sap.ui.core\\" xmlns:mvc=\\"sap.ui.core.mvc\\" xmlns=\\"sap.m\\"
	xmlns:html=\\"http://www.w3.org/1999/xhtml\\" controllerName=\\"my.test.App.ext.customPage.CustomPage\\">
	
	<App>
		<pages>
			<Page title=\\"CustomPage\\">
				<content></content>
			</Page>
		</pages>
	</App>
	
</mvc:View>",
    "state": "modified",
  },
  "webapp/manifest.json": Object {
    "contents": "{
  \\"sap.app\\": {
    \\"id\\": \\"my.test.App\\"
  },
  \\"sap.ui5\\": {
    \\"routing\\": {
      \\"routes\\": [
        {
          \\"name\\": \\"RootEnityCustomPage\\",
          \\"pattern\\": \\"RootEnity({key}):?query:\\",
          \\"target\\": \\"RootEnityCustomPage\\"
        }
      ],
      \\"targets\\": {
        \\"RootEnityCustomPage\\": {
          \\"type\\": \\"Component\\",
          \\"id\\": \\"my.test.App.ext.customPage\\",
          \\"name\\": \\"sap.fe.core.fpm\\",
          \\"options\\": {
            \\"settings\\": {
              \\"viewName\\": \\"my.test.App.ext.customPage.CustomPage\\",
              \\"entitySet\\": \\"RootEnity\\",
              \\"navigation\\": {}
            }
          }
        }
      }
    },
    \\"targets\\": {
      \\"TestObjectPage\\": {}
    }
  }
}
",
    "state": "modified",
  },
}
`;

exports[`CustomPage Add a custom page with navigation to it 1`] = `
Object {
  "sap.app": Object {
    "id": "my.test.App",
  },
  "sap.ui5": Object {
    "routing": Object {
      "routes": Array [
        Object {
          "name": "ChildEntityCustomPage",
          "pattern": "RootEntity({key})/navToChildEntity({key2}):?query:",
          "target": "ChildEntityCustomPage",
        },
      ],
      "targets": Object {
        "ChildEntityCustomPage": Object {
          "id": "my.test.App.ext.customPage",
          "name": "sap.fe.core.fpm",
          "options": Object {
            "settings": Object {
              "entitySet": "ChildEntity",
              "navigation": Object {},
              "viewName": "my.test.App.ext.customPage.CustomPage",
            },
          },
          "type": "Component",
        },
        "TestObjectPage": Object {
          "options": Object {
            "settings": Object {
              "navigation": Object {
                "navToChildEntity": Object {
                  "detail": Object {
                    "route": "ChildEntityCustomPage",
                  },
                },
              },
            },
          },
        },
      },
    },
    "targets": Object {
      "TestObjectPage": Object {},
    },
  },
}
`;
